src/server/shell/abstract_shell.cpp

```diff
@@ -406,102 +432,114 @@ void msh::AbstractShell::set_focus_to(
+        set_keyboard_focus_surface(lock, new_keyboard_focus_surface);
+    }
+
     update_focus_locked(lock, focus_session, focus_surface);
 }
 
-void msh::AbstractShell::notify_focus_locked(
-    std::unique_lock<std::mutex> const& /*lock*/,
-    std::shared_ptr<ms::Surface> const& new_focus_surface)
+void msh::AbstractShell::notify_active_surfaces(
+    std::unique_lock<std::mutex> const&,
+    std::shared_ptr<ms::Surface> const& new_keyboard_focus_surface,
+    std::vector<std::shared_ptr<ms::Surface>> const& new_active_surfaces)
 {
-    auto const current_focus = notified_focus_surface.lock();
+    std::vector<std::shared_ptr<ms::Surface>> prev_active_surfaces;
+    prev_active_surfaces.reserve(notified_active_surfaces.size());
 
-    if (current_focus != new_focus_surface)
+    for (auto const& current_active_weak: notified_active_surfaces)
     {
-        std::vector<std::shared_ptr<ms::Surface>> new_active_surfaces;
-        for (auto item = new_focus_surface; item; item = item->parent())
-        {
-            new_active_surfaces.insert(begin(new_active_surfaces), item);
-        }
-
-        std::vector<std::shared_ptr<ms::Surface>> current_focus_tree;
-
-        notified_focus_surface = new_focus_surface;
-        seat->reset_confinement_regions();
-
-        for (auto const& item : notified_active_surfaces)
+        if (auto const current_active = current_active_weak.lock())
         {
-            if (auto const active = item.lock())
+            prev_active_surfaces.push_back(current_active);
+            // If a surface that was previously active is not in the set of new active surfaces, notify it
+            if (find(begin(new_active_surfaces), end(new_active_surfaces), current_active) == end(new_active_surfaces))
             {
-                current_focus_tree.push_back(active);
-                if (find(begin(new_active_surfaces), end(new_active_surfaces), active) == end(new_active_surfaces))
-                {
-                    active->set_focus_state(mir_window_focus_state_unfocused);
-
-                    // When a menu loses focus we should close and unmap it
-                    if (active->type() == mir_window_type_menu || active->type() == mir_window_type_gloss)
-                    {
-                        active->request_client_surface_close();
-                        active->hide();
-                    }
-                }
-                else if (active == current_focus)
+                current_active->set_focus_state(mir_window_focus_state_unfocused);
+
+                // When a menu loses focus we should close and unmap it
+                if (current_active->type() == mir_window_type_menu || current_active->type() == mir_window_type_gloss)
                 {
-                    active->set_focus_state(mir_window_focus_state_active);
+                    current_active->request_client_surface_close();
+                    current_active->hide();
```

Comment:
Is this going to do the right thing? We need to not only close the popup, but also any parent popups it has, and in child-first order, whereas I *think* you're iterating over this in parent-first order?